File: l2/component/component.go
//go:build linux || darwin

package component

import (
	"plugin"
	"fmt"
	"os"
)

type Component *plugin.Plugin

func ReturnComponentFunction(_Component Component, Name string) any {
	function, err := (*plugin.Plugin)(_Component).Lookup(Name)
	if err != nil {
		fmt.Println("luna-l2: failed sending '" + Name + "' to component:", err)
		return function
	}
	return function
}

func InitializeComponent(Path string) Component {
	fmt.Println("luna-l2: initializing component with path", Path)
	_Component, err := plugin.Open(Path)
	if err != nil {
		fmt.Println("luna-l2: failed to initialize component with path '" + Path + "':", err)
		os.Exit(1)
	}
	return Component(_Component)
}

File: l2/component/component.c
#ifndef _WIN32

#include <dlfcn.h>
#include <stdio.h>
#include <stdlib.h>

void* return_component_function(void* handle, char* name) {
    void* function_handle = dlsym(handle, name);
    if (function_handle == NULL) {
        printf("luna-l2: failed to return function '%s': %s", name, dlerror());
    }
    return function_handle;
}

void* initialize_component(char* path) {
    void* handle = dlopen(path, RTLD_LAZY);
    if (handle == NULL) {
        printf("luna-l2: failed to initialize component with path '%s': %s\n", path, dlerror());
        exit(1);
    }

    return handle;
}


#endif
